contrib/submit-protocol/src/App/ProtocolForm.tsx

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) Microsoft Corporation
+// All rights reserved.
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import React from "react";
+import { initializeIcons } from "office-ui-fabric-react/lib/Icons";
+import { Fabric } from "office-ui-fabric-react/lib/Fabric";
+import { Label } from "office-ui-fabric-react/lib/Label";
+import { TextField } from "office-ui-fabric-react/lib/TextField";
+import { Panel, PanelType } from "office-ui-fabric-react/lib/Panel";
+import { DefaultButton, PrimaryButton } from "office-ui-fabric-react/lib/Button";
+import crypto from "crypto";
+import MonacoEditor from "react-monaco-editor";
+import jsyaml from "js-yaml";
+
+interface IProtocolProps {
+  api: string;
+  user: string;
+  token: string;
+  source ?: {
+    jobName: string;
+    user: string;
+  };
+}
+
+interface IProtocolState {
+  jobName: string;
+  protocol: {[key: string]: string | object};
+  protocolYAML: string;
+  showEditor: boolean;
+}
+
+export default class ProtocolForm extends React.Component<IProtocolProps, IProtocolState> {
+  constructor(props: IProtocolProps) {
+    super(props);
+    initializeIcons();
+
+    this.state = {
+      jobName: "",
+      protocol: Object.create(null),
+      protocolYAML: "",
+      showEditor: false,
+    };
+
+    const source = this.props.source;
+    if (source && source.jobName && source.user) {
+      window.fetch(
```

Comment:
Side effects (async http request) should be put in componentDidMount

Reply:
Updated.

Reply:
Fetch in constructor should be moved to componentDidMount as well...

Reply:
only move `fetch` from constructor to `componentDidMount`, which is called before `render`.

`fetch` in `submitProtocol` is called when submit button is clicked, so I removed from `componentDidMount`.

Reply:
added loading spinner.